hepcrawl/testlib/celery_monitor.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of hepcrawl.
+# Copyright (C) 2017 CERN.
+#
+# hepcrawl is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""Celery monitor dealing with celery tasks for functional tests."""
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+from itertools import islice
+
+
+class CeleryMonitor(object):
+    def __init__(self, app, monitor_timeout=3, monitor_iter_limit=100):
+        self.results = []
+        self.recv = None
+        self.app = app
+        self.connection = None
+        self.monitor_timeout = monitor_timeout
+        self.monitor_iter_limit = monitor_iter_limit
+
+    def __enter__(self):
+        state = self.app.events.State()
+
+        def announce_succeeded_tasks(event):
+            state.event(event)
+            task = state.tasks.get(event['uuid'])
+            print('TASK SUCCEEDED: %s[%s] %s' % (task.name, task.uuid, task.info(),))
+            tasks = self.app.AsyncResult(task.id)
+            for task in tasks.result:
+                self.results.append(task)
+            self.recv.should_stop = True
+
+        def announce_failed_tasks(event):
+            state.event(event)
+            task = state.tasks.get(event['uuid'])
+            print('TASK FAILED: %s[%s] %s' % (task.name, task.uuid, task.info(),))
+            self.results.append(task.info())
+            self.recv.should_stop = True
+
+        self.app.control.enable_events()
+        self.connection = self.app.connection()
+        self.recv = self.app.events.Receiver(self.connection, handlers={
+            'task-succeeded': announce_succeeded_tasks,
+            'task-failed': announce_failed_tasks,
+        })
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        events_iter = self.recv.itercapture(limit=None, timeout=self.monitor_timeout, wakeup=True)
+        self._wait_for_results(events_iter)
+        self.connection.__exit__()
+
+    def _wait_for_results(self, events_iter):
+        any(islice(events_iter, self.monitor_iter_limit))
+
+    @classmethod
+    def do_crawl(cls,
+                 app,
+                 monitor_timeout,
+                 monitor_iter_limit,
+                 crawler_instance,
+                 project='hepcrawl',
+                 spider='WSP',
+                 settings=None,
+                 **crawler_arguments):
+
+        if settings is None:
+            settings = {}
+
+        with cls(app, monitor_timeout=monitor_timeout, monitor_iter_limit=monitor_iter_limit) as my_monitor:
+            crawler_instance.schedule(
+                project=project,
+                spider=spider,
+                settings=settings or {},
```

Comment:
Because having logic inline with the parameter of a call is quite harder to debug than having the logic by itself.